pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -426,17 +428,25 @@ private PinotResourceManagerResponse uploadSegment(File indexDir, SegmentMetadat
       return response;
     }
 
-    final OfflineSegmentZKMetadata originalOfflineSegmentZkMetadata = new OfflineSegmentZKMetadata(znRecord);
+    OfflineSegmentZKMetadata originalOfflineSegmentZkMetadata = new OfflineSegmentZKMetadata(znRecord);
 
     long minionExpectedCrc = -1L;
     boolean isMinionConfigured = offlineTableConfig.getTaskConfig() != null;
-    boolean minionHeaderPresent = headers.getHeaderString(HttpHeaders.USER_AGENT).contains(CommonConstants.Minion.MINION_HEADER);
+    String userAgentHeader = headers.getHeaderString(HttpHeaders.USER_AGENT);
+    boolean minionHeaderPresent = userAgentHeader.contains(CommonConstants.Minion.MINION_HEADER_PREFIX);
+    String taskType = UserAgentHeader.getTaskType(userAgentHeader);
     boolean metadataRewritten = false;
+
     if (isMinionConfigured) {
       // Minion configured use case so that we write to zk sparingly
       String segmentPushStatus = originalOfflineSegmentZkMetadata.getSegmentPushStatus();
       SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd'T'HH:mm:ss.SSS");
 
+      if (taskType.equals(MinionConstants.PurgeTask.TASK_TYPE)) {
```

Comment:
I am not happy that we need to add checks for minion and task type. Can we come up with a better design?

Reply:
I think minion-specific logic can be good(reasons below), but maybe not task-specific logic. However, I couldn't think of a better way to do it. I also think that purge is a specific use case, and we won't necessarily have to do this for other tasks.

Minion has specific logic that keeps creation time the same. Currently we reject segments of creation time the same or less. Additionally, a future looking optimization is that we don't need to necessarily refresh the whole segment in the case of use cases where minion generates segments of the same crc ( we don't want to reject, we still want to change refresh time, but that can be an optimization, for example ).

Reply:
I dont see a case where minion would create a segment with same CRC, even in the rare case that it does, its ok to refresh it. 

can you please provide more details on the creation time. why does minion keep the creation time the same. Do we introduce a concept of update time?

Reply:
This can happen in the case of purge. We want to change metadata, eg: purge time, when a segment is purged so that we can monitor and schedule appropriately. 

The way minion is designed we decided to keep the creation time the same for all minion tasks. In all the optimization tasks, that is the plan as well. The idea is that minion is working on the "same segment," and we use creation time in order to denote that.

Reply:
Subbu and I discussed this offline, and we discussed storing a map of minionHistory, which will contain the task - timestamp. Thoughts? This way, we don't need task specific logic.

Reply:
minion history will be a hashmap of minion-task-name to timestamp. the timestamp will be the last time that particular task touched the segment.

This will mean a change to the existing sgement metadata (not named appropriately) that has been out for about a week, or add a new segment metadatra and render the old one unused.